File: ml-benchmark/synduce/constraints/empty_right_subtree/contains.f
Inductive Tree = nil Unit | leaf Int | node {Int, Tree, Tree};

size = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    nil _ -> 0
  | leaf _ -> 1
  | node {_, l, r} -> + 1 (+ (f l) (f r))
  end
);

empty_right = fix (
  \f: Tree -> Bool. \t: Tree.
  match t with
    node {_, l, r} -> and (== 0 (size r)) (f l)
  | _ -> true
  end
);

@Input w: Int;

spec = fix (
  \f: Tree -> Int. \t: Tree.
  match t with
    nil _ -> 0
  | leaf a -> if == a w then 1 else 0
  | node {a, l, r} ->
    if == a w then 1 else if == 1 (f l) then 1 else f r
  end
);

target = fix (
  \f: Tree -> Compress Tree. \t: Tree.
  match t with
    nil _ -> nil unit
  | leaf a -> leaf a
  | node {a, l, r} -> node {a, f l, r} /* Avoid the recursion on the right branch.*/
  end
);

main = \t: Tree. if empty_right t then spec (target t) else 0;

File: ml-benchmark/synduce/constraints/empty_right_subtree/contains.ml
type tree =
  | Nil of unit
  | Leaf of int
  | Node of int * tree * tree

let rec size t =
  match t with
  | Nil _ -> 0
  | Leaf _ -> 1
  | Node (_, l, r) -> 1 + size l + size r

let rec empty_right t =
  match t with
  | Node (_, l, r) -> (size r = 0) && empty_right l
  | _ -> true

let w = 0

let rec spec t =
  match t with
  | Nil _ -> 0
  | Leaf a -> if a = w then 1 else 0
  | Node (a, l, r) ->
      if a = w then 1 else if spec l = 1 then 1 else spec r

let rec target t =
  match t with
  | Nil _ -> Nil ()
  | Leaf a -> Leaf a
  | Node (a, l, r) -> Node (a, target l, r)

let program t =
  if empty_right t then spec (target t) else 0